File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

/*
 * 길이가 같은 큐 2개
 * 하나에서 뽑아 다른 하나에 삽입할 수 있다
 * 목표 : 두 큐의 합이 같게 만들 수 있는 최소 횟수
 */

int solution(vector<int> queue1, vector<int> queue2) {
    int answer = 0;
    vector<ll> mergedQueue;
    mergedQueue.reserve(queue1.size() + queue2.size()); // 메모리 할당 최적화
    mergedQueue.insert(mergedQueue.end(), queue1.begin(), queue1.end());
    mergedQueue.insert(mergedQueue.end(), queue2.begin(), queue2.end());

    int n = queue1.size();
    ll sum = 0;
    ll q1 = 0;
    ll q2 = 0;

    for(int i = 0; i<n; i++){
        q1 += queue1[i];
        q2 += queue2[i];
        sum += queue1[i] + queue2[i];
    }
    if(sum % 2 != 0) return -1;


    int l = 0;
    int r = n;
    bool possible = false;
    
    while(l<=r && r< mergedQueue.size()){
        if(q1 == q2) {
            possible = true;
            break;
        }
        if(q1 > q2){
            q1 -= mergedQueue[l];
            q2 += mergedQueue[l];
            l++;
            answer++;
        }else{
            q2 -= mergedQueue[r];
            q1 += mergedQueue[r];
            r++;
            answer++;
        }
    }


    return possible ? answer : -1;
}

File: 프로그래머스/2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = -2;
        
        long sum = 0;
        int limit = queue1.length + queue2.length;
        
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        long q1s = 0;
        long q2s = 0;
        
        for(int num : queue1){
            q1s += (long) num;
            sum += (long) num;
            q1.add(num);
        }
        
        for(int num : queue2){
            q2.add(num);
            sum += (long) num;
            q2s += (long) num;
        }
        
        
        if(sum % 2 != 0) return -1;
        int cnt = 0;
        
        while(limit*2 >= cnt){
            if(q1s == q2s){
                return cnt;
            }else if(q1s > q2s){
                int pop = q1.poll();
                q1s -= pop;
                q2s += pop;
                q2.add(pop);
                cnt++;
            }else {
                int pop = q2.poll();
                q1s += pop;
                q2s -= pop;
                q1.add(pop);
                cnt++;
            }
        }
        
        
        return -1;
    }
}